src/index.ts

```diff
@@ -1,44 +1,99 @@
+import defaultSessionProcessFunc from "./util/defaultSessionProcessFunc"
+import whenReady from "./util/loadWhenDocumentReady"
 
-export function init(settings: PageSettings): void {
-    loadWhenReady(() => {
-        console.log('Document loaded and ready')
-        settings = settings || DefaultPageSettings
+/**
+ * Called by tag owner code to initialize a RUM session, either immediately or
+ * after some delay.
+ *
+ * @remarks
+ * Waits for the page to load before processing.
+ *
+ * @param settings Specifies settings affecting client behavior. These are
+ * determined by the tag owner at runtime, so may be used to specify page-level
+ * overrides to general defaults.
+ */
+export default function init(settings: ClientSettings): Promise<SessionResult> {
+    return whenReady().then(() => {
         if (settings.preConfigStartDelay) {
-            setTimeout(startLater(settings), settings.preConfigStartDelay)
-        } else {
-            start(settings)
+            return startLater(settings.preConfigStartDelay, settings)
         }
+        return start(settings)
     })
 }
 
-function startLater(settings: PageSettings): () => void {
-    return function () {
-        start(settings)
-    }
+/**
+ * Called internally if a non-zero {@link ClientSettings.preConfigStartDelay}
+ * setting has been specified. Calls {@link start} after the delay.
+ *
+ * @param delay The approximate time to wait (in milliseconds).
+ * @param settings The settings object passed to {@link init}.
+ */
+function startLater(
+    delay: number,
+    settings: ClientSettings,
+): Promise<SessionResult> {
+    return new Promise((resolve) => {
+        setTimeout(() => {
+            start(settings).then((result) => resolve(result))
+        }, delay)
+    })
 }
 
-function start(settings: PageSettings): void {
-    const promises = settings.providers
-        .filter(p => p.testUserAgent())
-        .map(p => p.fetchConfig())
-    if (promises.length) {
-        Promise.all(promises)
-            .then(providerConfigs => {
-                const temp = <[Provider, unknown, unknown[]][]>providerConfigs
-                for (let config of temp) {
-                    let provider: Provider = config[0]
-                    for (let task of config[2]) {
-                        provider.executeTask(config[1], task)
-                    }
-                }
-            })
-            .catch(error => {
-                console.error(error)
-            })
-    } else {
-        console.log('No providers promised to make config calls')
-    }
+/**
+ * Called immediately by {@link init} if no
+ * {@link ClientSettings.preConfigStartDelay} setting has been specified.
+ *
+ * @param settings The settings object passed to {@link init}.
+ */
+function start(settings: ClientSettings): Promise<SessionResult> {
+    const activeProviders = settings.providers.filter((p) => p.shouldRun())
+    return Promise.allSettled(
```

Comment:
I know that I suggested using `allSettled` however I'm slightly concerned about it's browser support. So we should ensure it is sufficiently polyfill'ed. Which has made me realise that we're not using Babel yet in this project, which we should do! Not a blocker for this PR, something to discuss offline and I can happily follow up with a PR to add babel with appropriate plugins to polyfill etc.